Node repurpose tool docs

henningandersen opened the pull request:
Added documentation for node repurpose tool.

Follow-up to #39403

henningandersen committed:
Node repurpose tool docs

Added documentation for node repurpose tool.

elasticmachine commented:
Pinging @elastic/es-distributed

andrershov reviewed (changes requested):
As I mentioned earlier, there would be problems with combining documentation for unsafe-bootstrap/detach-cluster and repurpose. For example, when reading the description section there are several paragraphs that talk about unsafe cluster recover, but after that, there is a paragraph about repurpose and just one question arises - what is it all about?
I still think that the best idea is to explicitly have elasticsearch-repurpose tool separately.
But if we don't want to do it, we need to think carefully about grouping description/examples in such a way that do not confuse the reader.
The only idea that comes to my mind is to make too child pages for elasticsearch-node tool - one for unsafe-bootstrap/detach-cluster and another for repurpose. On the parent page, we can leave short general description and synopsis, but everything else comes to separate pages. But still, it looks artificial.
@ywelsch @DaveCTurner do you have better ideas?

henningandersen commented:
Thanks for looking into this @andrershov . I tend to think the tool is quite coherent in that all 3 commands relate to the node's role in the cluster and tries to avoid someone messing directly with files in the data path.
I could probably make that clearer in the documentation, but I was feeling a little constrained by not wanting to rewrite the entire intro section since you all spent a good deal of time ensuring the correctness of it, paying attention to many details in the original node-tool docs PR #37812.
Just my cents, will await feedback from @ywelsch and @DaveCTurner .

andrershov reviewed (commented):
@henningandersen By intro I mean the paragraph that is already in the documentation
```
The `elasticsearch-node` command enables you to perform unsafe operations that
risk data loss but which may help to recover some data in a disaster.	
```
This and synopsis can stay in the parent page - everything else will go to the child pages. And child page for unsafe-bootstrap/detach-cluster will look the same way it was before your changes. All your additions will go to the corresponding sections of repurpose child page.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -2,14 +2,15 @@
 The `elasticsearch-node` command enables you to perform unsafe operations that
-risk data loss but which may help to recover some data in a disaster.
+risk data loss but which may help to recover some data in a disaster and
+repurpose nodes into new roles.
```
Suggest breaking into two simpler sentences:

```
The `elasticsearch-node` command enables you to perform certain unsafe
operations on a node that are only possible while it is shut down. This command
allows you to adjust the <<modules-node,role>> of a node and may be able to
recover some data after a disaster.
```

Which reminds me, the docs about roles are at `docs/reference/modules/node.asciidoc` and I think we should have docs there to say that you now have to take special steps to change the role of a node.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -2,14 +2,15 @@
 [source,shell]
 --------------------------------------------------
-bin/elasticsearch-node unsafe-bootstrap|detach-cluster
+bin/elasticsearch-node unsafe-bootstrap|detach-cluster|repurpose
```
I'm inclined to put the `repurpose` command first here (and elsewhere) because of the three commands it's the one that we hope is the most used.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -53,19 +54,28 @@ way forward that does not risk data loss, but it may be possible to use the
 data from the failed cluster.
 
-This tool has two modes:
+If a node that previously held data or were master-eligible is repurposed into
```
I suggest:

1. moving this to the top of the `Description` section,
2. making new subheadings e.g. `==== Changing the roles of a node` and `==== Recovering data after a disaster` to separate the sections more clearly, and
3. expanding this paragraph into a longer description of the concepts (and/or describing the concepts in `docs/reference/modules/node.asciidoc` and linking to them).

henningandersen replied:
About item 3: I think master-eligible node and data node are already described in `docs/reference/modules/node.asciidoc`. I can certainly link to those descriptions. But it sounds like you wanted something more than that, can you elaborate on that?

DaveCTurner replied:
Sure, sorry, I'll clarify. Today the `Node` docs describe the `node.data` and `node.master` settings but do not say anything about how to go about changing them. This makes sense in 6.x because you don't have to do anything particularly special to change them, but in 7.0 this is no longer the case. I think that page needs to mention that you should only really set these settings before starting a node for the first time, and also talk about (or at least link to) the extra steps you need to do if you want to change them later.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -53,19 +54,28 @@ way forward that does not risk data loss, but it may be possible to use the
   This is known as _unsafe cluster bootstrapping_.
 
-* `elastisearch-node detach-cluster` enables you to move nodes from one cluster
+* `elasticsearch-node detach-cluster` enables you to move nodes from one cluster
```
D'oh :)

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -53,19 +54,28 @@ way forward that does not risk data loss, but it may be possible to use the
   possible, it also enables you to
   move nodes into a brand-new cluster.
 
+* `elasticsearch-node repurpose` reports on and performs necessary index meta-data
```
I'd put this mode first.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -206,6 +216,44 @@ The message `Node was successfully detached from the cluster` does not mean
 its job.
 
+
+[[node-tool-repurpose]]
```
I'd put this section above the other two.

henningandersen commented:
@andrershov by intro section, I really meant the `Description` section, I have updated my comment above to clarify that.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -206,6 +216,44 @@ The message `Node was successfully detached from the cluster` does not mean
+[float]
+==== Repurposing nodes
+
+A master-eligible and data carrying node holds on-disk index data:
```
I think I'd prefer to try and split this whole section into two parallel descriptions about `node.data` and `node.master` respectively, because here it's a bit unclear what this says about a master-ineligible data node or (conversely) a dedicated master node. It might be a little repetitive, but I'm ok with that.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -215,6 +263,8 @@ one-node cluster.
 `detach-cluster`:: Specifies to unsafely detach this node from its cluster so
 it can join a different cluster.
 
+`repurpose`:: Report on and cleanup excess persisted data.
```
I'd put this first.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -332,3 +382,45 @@ Confirm [y/N] y
 ----
 
+[float]
+==== Repurposing node as master, no-data
```
I'd put this first.

DaveCTurner reviewed (dismissed):
I'm inclined to agree with @andrershov that this change makes these docs feel a bit disjointed. See also [this comment](https://github.com/elastic/elasticsearch/pull/37812#issuecomment-471662475) with which I agree: eventually I think we want to move the conceptual overview of these things to a place that isn't buried so deeply. I do not think we should split the docs for each sub-command out into separate pages, although it's possible we would want to do this for the conceptual overview.

Yet I think it's important we have this tool documented here for now, and we can defer this larger refactoring until later.

I left some suggestions, mostly about structure/order, and also noted that I think there's something missing on the subject from `docs/reference/modules/node.asciidoc`.

henningandersen committed:
Node repurpose tool docs

Improved documentation for node repurpose tool.

henningandersen commented:
Thanks @andrershov and @DaveCTurner for your reviews. I tried to adhere to all comments in one commit: 3879859bafcada9cc42ad56095b7c673dc8e0e02 (felt unnatural to split into multiple commits). I did not split the page, but went for two subsections in the Description section.

Please have another look.

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
 === Examples
 
+[float]
+==== Repurposing node as master, no-data
```
```suggestion
==== Repurposing a data node as a dedicated master node
```

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
+----
+
+[float]
+==== Repurposing node as no-master, no-data
```
```suggestion
==== Repurposing a data node as a coordinating-only node
```

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
+[float]
+==== Repurposing node as no-master, no-data
+
+In this example, a node that previously held data is repurposed as no-master, no-data.
```
```suggestion
In this example, a node that previously held data is repurposed as a coordinating-only node.
```

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
+==== Repurposing node as no-master, no-data
+
+In this example, a node that previously held data is repurposed as no-master, no-data.
+First update the node settings to `node.master=true` and `node.data=false`. Then run the
```
```suggestion
First update the node's settings to `node.master: false` and `node.data: false` in its `elasticsearch.yml` config file. Then run the
```

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
+[float]
+==== Repurposing node as master, no-data
+
+In this example, a node that previously held data is repurposed as master, no-data.
```
```suggestion
In this example, a node that previously held data is repurposed as a dedicated master node.
```

DaveCTurner commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -230,6 +284,49 @@ to `0`, meaning to use the first node in the data path.
+==== Repurposing node as master, no-data
+
+In this example, a node that previously held data is repurposed as master, no-data.
+First update the node settings to `node.master=true` and `node.data=false`. Then run the
```
```suggestion
First update the node's settings to `node.master: true` and `node.data: false` in its `elasticsearch.yml` config file. Then run the
```

DaveCTurner reviewed (commented):
I left a few more nits but the structure feels good to me now. I would like to do a bit of wordsmithery with the longer sections of words but will defer that until after the weekend.

DaveCTurner committed:
Node repurpose tool docs

Clarifty that changes happen in .yml file.

Co-Authored-By: henningandersen <33268011+henningandersen@users.noreply.github.com>

DaveCTurner committed:
Node repurpose tool docs

Clarify that changes happen in .yml file.

Co-Authored-By: henningandersen <33268011+henningandersen@users.noreply.github.com>

henningandersen committed:
Node repurpose tool docs

Changed terminology a bit to use coordinating-only and dedicated master.
Though not entirely true, it gives a slightly easier read.

henningandersen commented:
Thanks @DaveCTurner , I have addressed the 4 changes, though not identically to what you requested. Let me know if you have additional comments.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool_docs

henningandersen commented:
@elasticmachine run elasticsearch-ci/1

DaveCTurner committed:
Rewording

DaveCTurner commented:
@henningandersen I have done a bit of work on this on my own branch (on top of this one): https://github.com/elastic/elasticsearch/compare/master...DaveCTurner:2019-04-08-feature_node_repurpose_tool_docs

I added more detail about repurposing to the `Node` page which simplified the tool docs a bit. Could you compare and bring in any changes you like?

henningandersen committed:
Node repurpose tool docs

Removed todo and changed metadata to data to also capture shard data.

henningandersen commented:
> I added more detail about repurposing to the `Node` page which simplified the tool docs a bit. Could you compare and bring in any changes you like?

Thanks @DaveCTurner , I took your changes as is and made two small changes only in 514831dfa6556f3f8435166c5537a3057ae6a962

DaveCTurner reviewed (approved):
LGTM

henningandersen committed:
Merge remote-tracking branch 'origin/master' into feature_node_repurpose_tool_docs

andrershov reviewed (approved):
LGTM2
